src/regex/dfa/dfa.py

```diff
@@ -30,14 +31,25 @@ def __call__(self, _: str):
 
 class CharacterClassMatcher(Callable[[str], int], Matcher):
     # TODO: Tests fail here
-    def __init__(self, character_class: list[CharacterRange], target_state: int):
+    def __init__(self, character_class: list[CharacterRange], is_negation: bool, target_state: int):
         self.character_class = character_class
+        self.is_negation = is_negation
         self.next_state = target_state
 
     def __call__(self, symbol: str) -> int | None:
-        for _range in self.character_class:
-            if _range.start <= ord(symbol) <= _range.end:
-                return self.next_state
+        for i, _range in enumerate(self.character_class):
+            is_in_range = _range.start <= ord(symbol) <= _range.end
+
+            match self.is_negation, is_in_range:
+                case True, True:
+                    return None
+                case True, False:
+                    if i == len(self.character_class) - 1:
+                        return self.next_state
+                    continue
+                case False, True:
+                    return self.next_state
+
```

Comment:
A more pythonic way using a comprehension (equivalent to a map) and using the `all` or `any` operation to reduce the list dependent on the `self.is_negation` flag.

```suggestion
        predicate: Callable[[CharacterRange], bool] = lambda char_range: (
            # != functions as XOR here
                self.is_negation != (char_range.start <= ord(symbol) <= char_range.end))

        # default case matches if at least one match exists
        if not self.is_negation and any(predicate(cr) for cr in self.character_class):
            return self.next_state
        # negated case matches if all are not matching (!∃: r <=> ∀: !r)
        elif self.is_negation and all(predicate(cr) for cr in self.character_class):
            return self.next_state
        else:
            return None
```